File: Backend/tshirt-store-backend/src/main/java/com/tshirtstore/service/AdminProductService.java
package com.tshirtstore.service;

import com.tshirtstore.dto.CreateVariantRequest;
import com.tshirtstore.dto.ProductResponse;
import com.tshirtstore.dto.admin.CreateProductRequest;
import com.tshirtstore.dto.admin.UpdateProductRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

public interface AdminProductService {

    ProductResponse createProduct(CreateProductRequest request);

    Page<ProductResponse> getAllProducts(Pageable pageable);

    ProductResponse getProductById(Long id);

    ProductResponse updateProduct(Long id, UpdateProductRequest request);

    void deleteProduct(Long id);

    ProductResponse addImageToProduct(Long productId, MultipartFile file, String altText, Integer displayOrder);

    void deleteProductImage(Long productId, Long imageId);

    ProductResponse addVariantToProduct(Long productId, CreateVariantRequest request);

    ProductResponse updateVariant(Long productId, Long variantId, CreateVariantRequest request);

    void deleteVariant(Long productId, Long variantId);
}

File: Frontend/T-Shirt/src/services/adminProductService.ts
import { api } from "./api";
import type { Product } from "../types/product";

export interface CreateVariantRequest {
  sku: string;
  color: string;
  colorCode?: string;
  size: string;
  price: number;
  compareAtPrice?: number;
  stock: number;
}

export interface CreateProductPayload {
  name: string;
  slug: string;
  description: string;
  category: string;
  price: number;
  compareAtPrice?: number;
  isNew?: boolean;
  isFeatured?: boolean;
  variants: CreateVariantRequest[];
}

export interface UpdateProductPayload {
  name?: string;
  slug?: string;
  description?: string;
  category?: string;
  price?: number;
  compareAtPrice?: number;
  isNew?: boolean;
  isFeatured?: boolean;
  active?: boolean;
  variants?: CreateVariantRequest[];
}

export const adminProductService = {
  getProducts: async (page = 0, size = 20) => {
    const response = await api.get(`/admin/products?page=${page}&size=${size}`);
    return response.data;
  },

  getProductById: async (id: number): Promise<Product> => {
    const response = await api.get(`/admin/products/${id}`);
    return response.data;
  },

  createProduct: async (payload: CreateProductPayload): Promise<Product> => {
    const response = await api.post("/admin/products", payload);
    return response.data;
  },

  updateProduct: async (id: number, payload: UpdateProductPayload): Promise<Product> => {
    const response = await api.put(`/admin/products/${id}`, payload);
    return response.data;
  },

  deleteProduct: async (id: number): Promise<void> => {
    await api.delete(`/admin/products/${id}`);
  },

  uploadImage: async (productId: number, file: File, altText?: string): Promise<Product> => {
    const formData = new FormData();
    formData.append("file", file);
    if (altText) formData.append("altText", altText);

    const response = await api.post(`/admin/products/${productId}/images`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });
    return response.data;
  },

  deleteImage: async (productId: number, imageId: number): Promise<void> => {
    await api.delete(`/admin/products/${productId}/images/${imageId}`);
  },

  addVariant: async (productId: number, variant: CreateVariantRequest): Promise<Product> => {
    const response = await api.post(`/admin/products/${productId}/variants`, variant);
    return response.data;
  },

  updateVariant: async (productId: number, variantId: number, variant: CreateVariantRequest): Promise<Product> => {
    const response = await api.put(`/admin/products/${productId}/variants/${variantId}`, variant);
    return response.data;
  },

  deleteVariant: async (productId: number, variantId: number): Promise<void> => {
    await api.delete(`/admin/products/${productId}/variants/${variantId}`);
  },
};
